### ISSUE TITLE
Incorrect scaling of image on HiDPI screen,, wayland, scale factor 2

### ISSUE THREAD

**Issue by voidastro4** [nomacs/nomacs#1126]
On Linux wlroots Sway (latest stable) Nomacs Wayland (NOT xwayland - confirmed)

```
swaymsg -t get_outputs
...
3840x2160 @ 60.000 Hz
Scale factor: 2.000000
Transform: normal
```

Nomacs image scaling is broken. Presumably due to the sway scaling factor, it incorrectly targets a 1080p resolution for scaling
This is known due to:
- Upscaling artifacts in the image on the 4K display
- Nomacs reported a % scale (in the corner) consistent with a 1080p screen instead of a 4K one.

(The images used in the test are greater than 4K in resolution)

The Nomacs UI is scaled correctly.

Workaround:
QT_SCALE_FACTOR=0.5

This fixes the image scaling. UI scaling is rolled back and is smaller than desirable.

I'm not aware of any other application that suffers this problem.
qView also uses QT and doesn't have this problem. I have since switched to it.

**Comment by Xenapte** [comment]
This issue is present on X11 too.

**Comment by fililip** [comment]
Happens on KDE Plasma 6.2 (Wayland) as well.

**Comment by fililip** [comment]
@scrubbbbs Setting `QT_FONT_DPI` together with `QT_SCALE_FACTOR` does indeed help with UI scaling, but it breaks the viewport (the image is not pixel-perfect anymore), at least with fractional scaling (eg. 125%). Do you have another workaround for that, or would it need a patch?

**Comment by scrubbbbs** [comment]
@fililip It seems like we need a way to disable wayland scaling on a per-application/per-window basis, or have an API or environment variable to tell sway/wayland that we are DPI-aware and it needs to leave us alone, but I cannot find any solution for that.

**Comment by z18d7** [comment]
How about adding font size selection in settings? There is already an option for icon size.

With `QT_SCALE_FACTOR=0.5`, custom font and icon sizes interface would look like somewhat close to what it supposed to look under HiDPI without affecting image scaling

**Comment by scrubbbbs** [comment]
It is [theoretically possible ](https://wayland-book.com/surfaces-in-depth/lifecycle.html) for a wayland window to be unscaled via the "scale" attribute of the surface protocol, I'm looking into the status of this in Qt/kwin.

**Comment by scrubbbbs** [comment]
> How about adding a font size selection to settings? There is already an option for icon size.

It's a possibility, but it won't fix the problem with incorrect image scaling.

**Comment by z18d7** [comment]
> > How about adding a font size selection to settings? There is already an option for icon size.
> It's a possibility, but it won't fix the problem with incorrect image scaling.

`QT_SCALE_FACTOR` is going to take care of that. And fonts twice as big as normal should significantly improve tiny UI.
Of course it's just an ugly workaround in case no other solutions are found.

**Comment by Trust2k** [comment]
I confirm the problem, also encountered it, switched to imv, but it is also not without drawbacks.

**Comment by voidastro4** [comment]
imv is a bad choice, it has no antialiasing capabilities. when images have higher resolution than your monitor or window the quality is horrible.

qView turned out better than nomacs in every way

**Comment by fililip** [comment]
~~I figured something out. Removing `setWorldMatrixEnabled(false)`, using `mImgViewRect` instead of `displayRect` and doing this in `DkBaseViewPort::draw(QPainter &painter, double opacity)` at the beginning of the function:~~

```cpp
qreal dpr = devicePixelRatioF();
QPointF center = rect().center();
painter.translate(center);
painter.scale(1.0f / dpr, 1.0f / dpr);
painter.translate(-center);
```

~~gives properly scaled results no matter the display scale set in KDE. Furthermore, dragging nomacs' window between mixed DPI screens scales the image accordingly.~~

~~This has an issue with images larger than the canvas though; when the app starts for the first time, an incorrect zoom level is established (2x less in my case) because QPainter is scaled down. But I think that is already better than what we have right now because images are correctly adjusted to display scale and zooming works as intended.~~

~~@scrubbbbs I'd be very happy to hear some suggestions as to how I should take this further, ie if there is a better way to propagate `devicePixelRatio`.~~

Ah, sorry, I just found #1102, that diff works wonders actually. Are you planning to revisit that idea by any chance?

**Comment by scrubbbbs** [comment]
> Ah, sorry, I just found [#1102](https://github.com/nomacs/nomacs/pull/1102), that diff works wonders actually. Are you planning to revisit that idea by any chance?

Yes as you've found out this is pretty much what #1102 does. The hard part is fixing other places affected by this simple change (save-as dialog, paint plugin, maybe printing, maybe frameless mode, sizing image to fit window, etc etc).

I've been thinking about alternate approaches and it seems like the most practical solution. I will probably work on this next once other PRs are merged. It's a feature I want and is one of the most voted issues now.

**Comment by scrubbbbs** [comment]
#1382 opened to work on this, could use some help to find all the issues.

**Comment by fililip** [comment]
> [#1382](https://github.com/nomacs/nomacs/pull/1382) opened to work on this, could use some help to find all the issues.

This looks to be in great shape on my system, though I'll admit I have not tested *everything* this PR implements yet, only the basic view and cropping. It's especially cool to see that the icons are now correctly rendered at 200%.

Thanks for working on this!

**Comment by scrubbbbs** [comment]
Fixed with #1382
